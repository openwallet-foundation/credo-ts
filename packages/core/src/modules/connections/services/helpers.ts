import type { Routing } from './ConnectionService'
import type { AgentContext } from '../../../agent'
import type { ResolvedDidCommService } from '../../didcomm'
import type { DidDocument, PeerDidNumAlgo } from '../../dids'
import type { DidDoc, PublicKey } from '../models'

import { Key, KeyType } from '../../../crypto'
import { CredoError } from '../../../error'
import {
  IndyAgentService,
  DidCommV1Service,
  DidDocumentBuilder,
  getEd25519VerificationKey2018,
  DidRepository,
  DidsApi,
  createPeerDidDocumentFromServices,
  DidDocumentRole,
} from '../../dids'
import { didDocumentJsonToNumAlgo1Did } from '../../dids/methods/peer/peerDidNumAlgo1'
import { EmbeddedAuthentication } from '../models'

export function convertToNewDidDocument(didDoc: DidDoc): DidDocument {
  const didDocumentBuilder = new DidDocumentBuilder('')

  const oldIdNewIdMapping: { [key: string]: string } = {}

  didDoc.authentication.forEach((auth) => {
    const { publicKey: pk } = auth

    // did:peer did documents can only use referenced keys.
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addAuthentication(ed25519VerificationMethod.id)

      // Only the auth is embedded, we also need to add the key to the verificationMethod
      // for referenced authentication this should already be the case
      if (auth instanceof EmbeddedAuthentication) {
        didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
      }
    }
  })

  didDoc.publicKey.forEach((pk) => {
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
    }
  })

  // FIXME: we reverse the didCommServices here, as the previous implementation was wrong
  // and we need to keep the same order to not break the did creation process.
  // When we implement the migration to did:peer:2 and did:peer:3 according to the
  // RFCs we can change it.
  didDoc.didCommServices.reverse().forEach((service) => {
    const serviceId = normalizeId(service.id)

    // For didcommv1, we need to replace the old id with the new ones
    if (service instanceof DidCommV1Service) {
      const recipientKeys = service.recipientKeys.map((keyId) => {
        const oldKeyId = normalizeId(keyId)
        return oldIdNewIdMapping[oldKeyId]
      })

      service = new DidCommV1Service({
        id: serviceId,
        recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        accept: service.accept,
        priority: service.priority,
      })
    } else if (service instanceof IndyAgentService) {
      service = new IndyAgentService({
        id: serviceId,
        recipientKeys: service.recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        priority: service.priority,
      })
    }

    didDocumentBuilder.addService(service)
  })

  const didDocument = didDocumentBuilder.build()

  const peerDid = didDocumentJsonToNumAlgo1Did(didDocument.toJSON())
  didDocument.id = peerDid

  return didDocument
}

function normalizeId(fullId: string): `#${string}` {
  // Some old dids use `;` as the delimiter for the id. If we can't find a `#`
  // and a `;` exists, we will parse everything after `;` as the id.
  if (!fullId.includes('#') && fullId.includes(';')) {
    const [, ...ids] = fullId.split(';')

    return `#${ids.join(';')}`
  }

  const [, ...ids] = fullId.split('#')
  return `#${ids.length ? ids.join('#') : fullId}`
}

function convertPublicKeyToVerificationMethod(publicKey: PublicKey) {
  if (!publicKey.value) {
    throw new CredoError(`Public key ${publicKey.id} does not have value property`)
  }
  const publicKeyBase58 = publicKey.value
  const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
  return getEd25519VerificationKey2018({
    id: `#${publicKeyBase58.slice(0, 8)}`,
    key: ed25519Key,
    controller: '#id',
  })
}

export function routingToServices(routing: Routing): ResolvedDidCommService[] {
  return routing.endpoints.map((endpoint, index) => ({
    id: `#inline-${index}`,
    serviceEndpoint: endpoint,
    recipientKeys: [routing.recipientKey],
    routingKeys: routing.routingKeys,
  }))
}

export async function getDidDocumentForCreatedDid(agentContext: AgentContext, did: string) {
  const didRecord = await agentContext.dependencyManager.resolve(DidRepository).findCreatedDid(agentContext, did)

  if (!didRecord?.didDocument) {
    throw new CredoError(`Could not get DidDocument for created did ${did}`)
  }
  return didRecord.didDocument
}

/**
 * Asserts that the keys we are going to use for creating a did document haven't already been used in another did document
 * Due to how DIDComm v1 works (only reference the key not the did in encrypted message) we can't have multiple dids containing
 * the same key as we won't know which did (and thus which connection) a message is intended for.
 */
export async function assertNoCreatedDidExistsForKeys(agentContext: AgentContext, recipientKeys: Key[]) {
  const didRepository = agentContext.dependencyManager.resolve(DidRepository)
  const recipientKeyFingerprints = recipientKeys.map((key) => key.fingerprint)

  const didsForServices = await didRepository.findByQuery(agentContext, {
    role: DidDocumentRole.Created,

    // We want an $or query so we query for each key individually, not one did document
    // containing exactly the same keys as the did document we are trying to create
    $or: recipientKeyFingerprints.map((fingerprint) => ({
      recipientKeyFingerprints: [fingerprint],
    })),
  })

  if (didsForServices.length > 0) {
    const allDidRecipientKeys = didsForServices.flatMap((did) => did.getTags().recipientKeyFingerprints ?? [])
    const matchingFingerprints = allDidRecipientKeys.filter((f) => recipientKeyFingerprints.includes(f))
    throw new CredoError(
      `A did already exists for some of the keys in the provided services. DIDComm v1 uses key based routing, and therefore it is not allowed to re-use the same key in multiple did documents for DIDComm. If you use the same 'routing' object for multiple invitations, instead provide an 'invitationDid' to the create invitation method. The following fingerprints are already in use: ${matchingFingerprints.join(
        ','
      )}`
    )
  }
}

export async function createPeerDidFromServices(
  agentContext: AgentContext,
  services: ResolvedDidCommService[],
  numAlgo: PeerDidNumAlgo
) {
  const didsApi = agentContext.dependencyManager.resolve(DidsApi)

  // Create did document without the id property
  const didDocument = createPeerDidDocumentFromServices(services)

  // Assert that the keys we are going to use for creating a did document haven't already been used in another did document
  await assertNoCreatedDidExistsForKeys(agentContext, didDocument.recipientKeys)

  // Register did:peer document. This will generate the id property and save it to a did record
  const result = await didsApi.create({
    method: 'peer',
    didDocument,
    options: {
      numAlgo,
    },
  })

  if (result.didState?.state !== 'finished') {
    throw new CredoError(`Did document creation failed: ${JSON.stringify(result.didState)}`)
  }

  return result.didState.didDocument
}
