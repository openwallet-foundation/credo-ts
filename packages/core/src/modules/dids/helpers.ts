import { Key, KeyType } from '../../crypto'
import { isDid } from '../../utils'

import { DidKey } from './methods/key'

export function isDidKey(key: string) {
  return isDid(key, 'key')
}

export function didKeyToVerkey(key: string) {
  if (isDidKey(key)) {
    const publicKeyBase58 = DidKey.fromDid(key).key.publicKeyBase58
    return publicKeyBase58
  }
  return key
}

export function verkeyToDidKey(key: string) {
  if (isDidKey(key)) return key
  const publicKeyBase58 = key
  const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
  const didKey = new DidKey(ed25519Key)
  return didKey.did
}

export function didKeyToInstanceOfKey(key: string) {
  const didKey = DidKey.fromDid(key)
  return didKey.key
}

export function verkeyToInstanceOfKey(verkey: string) {
  const ed25519Key = Key.fromPublicKeyBase58(verkey, KeyType.Ed25519)
  return ed25519Key
}
