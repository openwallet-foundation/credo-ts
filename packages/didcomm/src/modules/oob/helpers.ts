import type { OutOfBandDidCommService } from './domain/OutOfBandDidCommService'

import {
  DidCommV1Service,
  DidDocumentBuilder,
  DidKey,
  didDocumentToNumAlgo2Did,
  createPeerDidDocumentFromServices,
  didKeyToInstanceOfKey,
} from '@credo-ts/core'

// This method is kept to support searching for existing connections created by
// credo-ts <= 0.5.1
// TODO: Remove in 0.6.0 (when ConnectionRecord.invitationDid will be migrated)
export function outOfBandServiceToInlineKeysNumAlgo2Did(service: OutOfBandDidCommService) {
  const didDocument = new DidDocumentBuilder('')
    .addService(
      new DidCommV1Service({
        id: service.id,
        serviceEndpoint: service.serviceEndpoint,
        accept: service.accept,
        recipientKeys: service.recipientKeys.map((recipientKey) => {
          const did = DidKey.fromDid(recipientKey)
          return `${did.did}#${did.key.fingerprint}`
        }),
        // Map did:key:xxx to actual did:key:xxx#123
        routingKeys: service.routingKeys?.map((routingKey) => {
          const did = DidKey.fromDid(routingKey)
          return `${did.did}#${did.key.fingerprint}`
        }),
      })
    )
    .build()

  const did = didDocumentToNumAlgo2Did(didDocument)

  return did
}

export function outOfBandServiceToNumAlgo2Did(service: OutOfBandDidCommService) {
  const didDocument = createPeerDidDocumentFromServices([
    {
      id: service.id,
      recipientKeys: service.recipientKeys.map(didKeyToInstanceOfKey),
      serviceEndpoint: service.serviceEndpoint,
      routingKeys: service.routingKeys?.map(didKeyToInstanceOfKey) ?? [],
    },
  ])

  const did = didDocumentToNumAlgo2Did(didDocument)

  return did
}
