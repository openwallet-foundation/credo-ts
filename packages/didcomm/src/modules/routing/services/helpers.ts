import type { AgentContext, DidDocument } from '@credo-ts/core'

import { MediationRecipientService } from './MediationRecipientService'

export async function getMediationRecordForDidDocument(agentContext: AgentContext, didDocument: DidDocument) {
  const [mediatorRecord] = await agentContext.dependencyManager
    .resolve(MediationRecipientService)
    .findAllMediatorsByQuery(agentContext, {
      recipientKeys: didDocument.recipientKeys.map((key) => key.publicKeyBase58),
    })
  return mediatorRecord
}
