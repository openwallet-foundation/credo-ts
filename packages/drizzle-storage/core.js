module.exports = require('./build/core/bundle.js')
