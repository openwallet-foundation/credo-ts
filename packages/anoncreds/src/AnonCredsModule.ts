import type { DependencyManager, Module, Update } from '@credo-ts/core'
import type { AnonCredsModuleConfigOptions } from './AnonCredsModuleConfig'

import { AnonCredsDataIntegrityServiceSymbol } from '@credo-ts/core'

import { AnonCredsApi } from './AnonCredsApi'
import { AnonCredsModuleConfig } from './AnonCredsModuleConfig'
import { AnonCredsRsHolderService, AnonCredsRsIssuerService, AnonCredsRsVerifierService } from './anoncreds-rs'
import { AnonCredsDataIntegrityService } from './anoncreds-rs/AnonCredsDataIntegrityService'
import {
  AnonCredsCredentialDefinitionPrivateRepository,
  AnonCredsKeyCorrectnessProofRepository,
  AnonCredsLinkSecretRepository,
  AnonCredsRevocationRegistryDefinitionPrivateRepository,
  AnonCredsRevocationRegistryDefinitionRepository,
} from './repository'
import { AnonCredsCredentialDefinitionRepository } from './repository/AnonCredsCredentialDefinitionRepository'
import { AnonCredsSchemaRepository } from './repository/AnonCredsSchemaRepository'
import { AnonCredsHolderServiceSymbol, AnonCredsIssuerServiceSymbol, AnonCredsVerifierServiceSymbol } from './services'
import { AnonCredsRegistryService } from './services/registry/AnonCredsRegistryService'
import { updateAnonCredsModuleV0_3_1ToV0_4 } from './updates/0.3.1-0.4'
import { updateAnonCredsModuleV0_4ToV0_5 } from './updates/0.4-0.5'

/**
 * @public
 */
export class AnonCredsModule implements Module {
  public readonly config: AnonCredsModuleConfig
  public api = AnonCredsApi

  public constructor(config: AnonCredsModuleConfigOptions) {
    this.config = new AnonCredsModuleConfig(config)
  }

  public register(dependencyManager: DependencyManager) {
    // Config
    dependencyManager.registerInstance(AnonCredsModuleConfig, this.config)

    dependencyManager.registerSingleton(AnonCredsRegistryService)

    // Repositories
    dependencyManager.registerSingleton(AnonCredsSchemaRepository)
    dependencyManager.registerSingleton(AnonCredsCredentialDefinitionRepository)
    dependencyManager.registerSingleton(AnonCredsCredentialDefinitionPrivateRepository)
    dependencyManager.registerSingleton(AnonCredsKeyCorrectnessProofRepository)
    dependencyManager.registerSingleton(AnonCredsLinkSecretRepository)
    dependencyManager.registerSingleton(AnonCredsRevocationRegistryDefinitionRepository)
    dependencyManager.registerSingleton(AnonCredsRevocationRegistryDefinitionPrivateRepository)

    // TODO: should we allow to override the service?
    dependencyManager.registerSingleton(AnonCredsHolderServiceSymbol, AnonCredsRsHolderService)
    dependencyManager.registerSingleton(AnonCredsIssuerServiceSymbol, AnonCredsRsIssuerService)
    dependencyManager.registerSingleton(AnonCredsVerifierServiceSymbol, AnonCredsRsVerifierService)

    dependencyManager.registerSingleton(AnonCredsDataIntegrityServiceSymbol, AnonCredsDataIntegrityService)
  }

  public updates = [
    {
      fromVersion: '0.3.1',
      toVersion: '0.4',
      doUpdate: updateAnonCredsModuleV0_3_1ToV0_4,
    },
    {
      fromVersion: '0.4',
      toVersion: '0.5',
      doUpdate: updateAnonCredsModuleV0_4ToV0_5,
    },
  ] satisfies Update[]
}
