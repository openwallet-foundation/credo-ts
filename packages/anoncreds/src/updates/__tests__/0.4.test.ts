import { readFileSync } from 'fs'
import path from 'path'
import {
  Agent,
  CacheModule,
  DependencyManager,
  InMemoryLruCache,
  InjectionSymbols,
  UpdateAssistant,
  W3cCredentialRecord,
} from '@credo-ts/core'

import { InMemoryStorageService } from '../../../../../tests/InMemoryStorageService'
import { agentDependencies } from '../../../../core/tests'
import { InMemoryAnonCredsRegistry } from '../../../tests/InMemoryAnonCredsRegistry'
import { anoncreds } from '../../../tests/helpers'
import { AnonCredsModule } from '../../AnonCredsModule'
import {
  AnonCredsHolderServiceSymbol,
  AnonCredsIssuerServiceSymbol,
  AnonCredsVerifierServiceSymbol,
} from '../../services'

// Backup date / time is the unique identifier for a backup, needs to be unique for every test
const backupDate = new Date('2024-02-28T22:50:20.522Z')
jest.useFakeTimers().setSystemTime(backupDate)

// We need to mock the uuid generation to make sure we generate consistent uuids for the new records created.
let uuidCounter = 1
jest.mock('../../../../core/src/utils/uuid', () => {
  return {
    uuid: jest.fn().mockImplementation(() => `${uuidCounter++}-4e4f-41d9-94c4-f49351b811f1`),
  }
})

describe('UpdateAssistant | AnonCreds | v0.4 - v0.5', () => {
  it('should correctly update the credential exchange records for holders', async () => {
    const holderRecordsString = readFileSync(
      path.join(__dirname, '__fixtures__/holder-anoncreds-2-anoncreds-records.0.4.json'),
      'utf8'
    )

    const dependencyManager = new DependencyManager()
    const storageService = new InMemoryStorageService()
    dependencyManager.registerInstance(InjectionSymbols.StorageService, storageService)
    dependencyManager.registerInstance(AnonCredsIssuerServiceSymbol, {})
    dependencyManager.registerInstance(AnonCredsHolderServiceSymbol, {})
    dependencyManager.registerInstance(AnonCredsVerifierServiceSymbol, {})

    const agent = new Agent(
      {
        config: {
          label: 'Test Agent',
        },
        dependencies: agentDependencies,
        modules: {
          cache: new CacheModule({
            cache: new InMemoryLruCache({ limit: 10 }),
          }),
          // We need to include the AnonCredsModule to run the updates
          anoncreds: new AnonCredsModule({
            registries: [new InMemoryAnonCredsRegistry()],
            anoncreds,
          }),
        },
      },
      dependencyManager
    )

    const updateAssistant = new UpdateAssistant(agent, {
      v0_1ToV0_2: {
        mediationRoleUpdateStrategy: 'doNotChange',
      },
    })

    await updateAssistant.initialize()

    // Set storage after initialization. This mimics as if this wallet
    // is opened as an existing wallet instead of a new wallet
    storageService.contextCorrelationIdToRecords = {
      default: {
        records: JSON.parse(holderRecordsString),
        creationDate: new Date(),
      },
    }

    expect(await updateAssistant.isUpToDate('0.5')).toBe(false)
    expect(await updateAssistant.getNeededUpdates('0.5')).toEqual([
      {
        fromVersion: '0.4',
        toVersion: '0.5',
        doUpdate: expect.any(Function),
      },
    ])

    await updateAssistant.update({ updateToVersion: '0.5' })

    expect(await updateAssistant.isUpToDate('0.5')).toBe(true)
    expect(await updateAssistant.getNeededUpdates('0.5')).toEqual([])

    // We mock the system time, however the issuanceDate is set by AnonCreds RS in rust, so we need to
    // manually set the issuanceDate to the current date (which is mocked) to not have inconsistent snapshot
    for (const record of Object.values(
      storageService.contextCorrelationIdToRecords[agent.context.contextCorrelationId].records
    )) {
      if (record.type !== W3cCredentialRecord.type) continue

      // biome-ignore lint/suspicious/noExplicitAny: <explanation>
      const recordValue = record.value as any
      recordValue.credential.issuanceDate = new Date()
    }

    expect(storageService.contextCorrelationIdToRecords[agent.context.contextCorrelationId].records).toMatchSnapshot()

    await agent.shutdown()
  })
})
