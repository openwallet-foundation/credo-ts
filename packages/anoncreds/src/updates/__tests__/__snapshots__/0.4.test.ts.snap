// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAssistant | AnonCreds | v0.4 - v0.5 should correctly update the credential exchange records for holders 1`] = `
{
  "1-4e4f-41d9-94c4-f49351b811f1": {
    "id": "1-4e4f-41d9-94c4-f49351b811f1",
    "tags": {
      "anonCredsAttr::test::marker": true,
      "anonCredsAttr::test::value": "test",
      "anonCredsCredentialDefinitionId": "did:indy:bcovrin:test:6LHqdUeWDWsL94zRc1ULEx/anoncreds/v0/CLAIM_DEF/400832/test",
      "anonCredsCredentialRevocationId": undefined,
      "anonCredsLinkSecretId": "ae343b1d-e2af-4706-9aae-2010a7f2c882",
      "anonCredsMethodName": "indy",
      "anonCredsRevocationRegistryId": undefined,
      "anonCredsSchemaId": "did:indy:bcovrin:test:6LHqdUeWDWsL94zRc1ULEx/anoncreds/v0/SCHEMA/test0.1599221872308001/1.0",
      "anonCredsSchemaIssuerId": "did:indy:bcovrin:test:6LHqdUeWDWsL94zRc1ULEx",
      "anonCredsSchemaName": "test0.1599221872308001",
      "anonCredsSchemaVersion": "1.0",
      "anonCredsUnqualifiedCredentialDefinitionId": "6LHqdUeWDWsL94zRc1ULEx:3:CL:400832:test",
      "anonCredsUnqualifiedIssuerId": "6LHqdUeWDWsL94zRc1ULEx",
      "anonCredsUnqualifiedRevocationRegistryId": undefined,
      "anonCredsUnqualifiedSchemaId": "6LHqdUeWDWsL94zRc1ULEx:2:test0.1599221872308001:1.0",
      "anonCredsUnqualifiedSchemaIssuerId": "6LHqdUeWDWsL94zRc1ULEx",
      "claimFormat": "ldp_vc",
      "contexts": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/data-integrity/v2",
      ],
      "cryptosuites": [
        "anoncreds-2023",
      ],
      "expandedTypes": [
        "https://www.w3.org/2018/credentials#VerifiableCredential",
      ],
      "givenId": undefined,
      "issuerId": "did:indy:bcovrin:test:6LHqdUeWDWsL94zRc1ULEx",
      "proofTypes": [
        "DataIntegrityProof",
      ],
      "schemaIds": [],
      "subjectIds": [],
      "types": [
        "VerifiableCredential",
      ],
    },
    "type": "W3cCredentialRecord",
    "value": {
      "createdAt": "2024-02-28T22:50:20.522Z",
      "credential": {
        "@context": [
          "https://www.w3.org/2018/credentials/v1",
          "https://w3id.org/security/data-integrity/v2",
          {
            "@vocab": "https://www.w3.org/ns/credentials/issuer-dependent#",
          },
        ],
        "credentialSubject": {
          "test": "test",
        },
        "issuanceDate": 2024-02-28T22:50:20.522Z,
        "issuer": "did:indy:bcovrin:test:6LHqdUeWDWsL94zRc1ULEx",
        "proof": [
          {
            "cryptosuite": "anoncreds-2023",
            "proofPurpose": "assertionMethod",
            "proofValue": "ukgGEqXNjaGVtYV9pZNlbZGlkOmluZHk6YmNvdnJpbjp0ZXN0OjZMSHFkVWVXRFdzTDk0elJjMVVMRXgvYW5vbmNyZWRzL3YwL1NDSEVNQS90ZXN0MC4xNTk5MjIxODcyMzA4MDAxLzEuMKtjcmVkX2RlZl9pZNlPZGlkOmluZHk6YmNvdnJpbjp0ZXN0OjZMSHFkVWVXRFdzTDk0elJjMVVMRXgvYW5vbmNyZWRzL3YwL0NMQUlNX0RFRi80MDA4MzIvdGVzdKlzaWduYXR1cmWCrHBfY3JlZGVudGlhbISjbV8y3AAgzL1rQBvM-szpzMImzNrMgcylCi7M8syxzKEQzLBBzKNTzOTM48zbzIlaNDzM1F91MaFh3AEAzOZYzMR6USQgEHTMoRbMrsydzKtnXBAGzITM7C4kPczaWG_Mh2nM0czezItlzLcnPMzVzIJSan06zMDMolbM48z3e8ynzOnMm8y6zOLMqMyizM3M-MyvzNnMrB3MtsyUc0ciO8yYK8ztzIHM4lg1PMyezJ7M0MyYecyrEsz4zOXMvldEF8yZIszbzNPMlT4xzIhfzMYCGA7M9cy3SczmzM3Mql0eScyJZ2UKKHRKGcy6zN8ozLvMvlLM_sy3WTjM_gfMrMzQMszKzKTMxDY2F8zmTWTM3EHMlwB4zLYAEsyeZcyIZBTMtlrMpczfSVdMzN1_zPQMzJ0QEAfM9MzIzIrMwsyeZ8zDH8zPzIRjUi42zJfM5wITADQlzMLM5grM8wrM-MzbC8zUB8ypzOPMj8yscMz7zMNbHAxCLQkpMER2zO7Ml3nMoB4yzI1GzJ7MkG3M61XM21jMhszZzLAvzJBgzK_M1MzlzMnM8gTMs8yCXQNdzLtuzL3MjMzYQsyDoWXcAEsQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABSMsydzL0HzPpfzJTM8CLM4czlJsyyzLmhdtwBVQ_M-EvM1cyczK3MqcyccwMqzIvM6kzM6UdwOcyDzPM1zN_Mnzt1zLvM2MyrIHzMoszPzOXMpMzhzJsBMMz5aszuzMbM38yrzNV3WMyFSczMzINpMMyOAszZOszaSMy1zIYBzKvMxHNuzKHM58zuzIvMvcyvzPkfFMyqzN_Mi1nMxADM1szOODERcsz3zJvMtcyCL0dUzJjMzMzwIsy8zJPMxMz_zMbMzcyWGyDM-S5FzOLM5szzYszbTczyLA8bMgAmRczCGHXMmwFhGQfMilMjHVBWcFXM28yWzOjMsMyJzKETHQsezOfMhcybesz7zKXM6wxwbsz8RszHMGBFzObM_mfMq8zWbMzpzKdDzOEwdizM0czeNwRWCR0DzM41bMzNzJczzOczzP7MgUErzK9cAhkDcg1GzLl7IszBPUBua1vMyCfMtszUzIvM6QcDVlzM_szUGczAzObMvszrzLHMl8yWeC0-KlDM7cy4VUPMsD_MmzjM-cz_zPnM0sz3zJPMvmRGzKnMqSdDzJnMpcybf3VnzL_MkMzazK4qI8ySzOjMnszHzMhTFWRozIbM7G1sMD3MjQrMucz4zPDMnMzNIMyPzKIfzJR_TcyMzNjMrVrMycyaKsy6zLjMlSEtKsy4zNfMsMyrY8y0zN5bVFXMgMzAzIzM4hrM2ylyzPJvD8yWVsyRFqxyX2NyZWRlbnRpYWzAu3NpZ25hdHVyZV9jb3JyZWN0bmVzc19wcm9vZoKic2XcAQDMqiPMi8zrzOPMgcy3zOo9GMz4zOlEKUjM3My2zNgAR8yZzJnM4D5fzOLM5szeEGnMscy6zN8mDszrzL3MrszfTHrM7MzBdMyvMcyizIXM3UE4asyXRSJnzK3MygZuXncNzLPMgWZIXQQPHMy3JcyCYF_MhDNnYCsMLl8LzLfM98yhGnZfzPfMrsygzKEAzI_M50nM28yxzLnMpFJ_ScyjEWfMiio2zIfMxszLzLBrIEFJD8zbzMEDPjDMiszIzKcTbCpXT3hzLiIld8zDzLbMyMyqHV3MzHnMwszzdCQEMlpGzOMRzJwCE8zizNXM4HnM8kYSzLcyzNNYzNrM7GbM1xTMwszkzKt5zNjM92QgzOZuCMzlzKQIRhZszP_MpMzbzLXM1sy0NyIHIszyzJBlzMLMjBrM4RI2Q8z2MMyRLRfM7iNkIMzEzORbzPkEesy5zJnMoMy5zMt7zMxYzLjM7zzMoC7Mt8y7OSNtHx84zI_M5m3MgszxM6Fj3AAgzJ49zLXMxljMmGvMqsy1BD3MpszXzMfM7MztzIwBQMyAdMzJCVXM38y0Y8ybzMhoeMyJ",
            "type": "DataIntegrityProof",
            "verificationMethod": "did:indy:bcovrin:test:6LHqdUeWDWsL94zRc1ULEx/anoncreds/v0/CLAIM_DEF/400832/test",
          },
        ],
        "type": [
          "VerifiableCredential",
        ],
      },
      "id": "1-4e4f-41d9-94c4-f49351b811f1",
      "metadata": {
        "_w3c/anonCredsMetadata": {
          "credentialRevocationId": undefined,
          "linkSecretId": "ae343b1d-e2af-4706-9aae-2010a7f2c882",
          "methodName": "indy",
        },
      },
      "updatedAt": "2024-02-28T22:50:20.522Z",
    },
  },
  "14720230-54b4-4735-8072-0e1902982429": {
    "id": "bb30d3ee-651a-4ffc-b1d9-30e4df23a045",
    "tags": {
      "attr": [
        "",
      ],
      "attr::test::marker": true,
      "attr::test::value": "test",
      "credentialDefinitionId": "6LHqdUeWDWsL94zRc1ULEx:3:CL:400832:test",
      "credentialId": "e1548638-fbe6-4e26-96fc-d536d453b380",
      "issuerId": "6LHqdUeWDWsL94zRc1ULEx",
      "linkSecretId": "ae343b1d-e2af-4706-9aae-2010a7f2c882",
      "methodName": "indy",
      "revocationRegistryId": null,
      "schemaId": "6LHqdUeWDWsL94zRc1ULEx:2:test0.1599221872308001:1.0",
      "schemaIssuerId": "6LHqdUeWDWsL94zRc1ULEx",
      "schemaName": "test0.1599221872308001",
      "schemaVersion": "1.0",
    },
    "type": "AnonCredsCredentialRecord",
    "value": {
      "credential": {
        "cred_def_id": "6LHqdUeWDWsL94zRc1ULEx:3:CL:400832:test",
        "rev_reg": null,
        "rev_reg_id": null,
        "schema_id": "6LHqdUeWDWsL94zRc1ULEx:2:test0.1599221872308001:1.0",
        "signature": {
          "p_credential": {
            "a": "29078583023644723417256963951834769429544893561312608041999039725779045478957665610539345669249033643845871310638205359827913345539893604784172221177874151867323789783352069858114677062382329201941336114640094448982492548621745418140287523358455796985452292603814050613022499206350481966450949147798607248367393022400871389441985483251995666110068431356102420069450232712370410071038529237338590022255675451148362600727584994665177285853081799683623674291603640314733666468918666175103622831604402215322289509594157875212406584558973682327127532803373424575230946721521968680075608178586165565162932643378236428403331",
            "e": "259344723055062059907025491480697571938277889515152306249728583105665800713306759149981690559193987143012367913206299323899696942213235956742930103927082934306204189453914854568633",
            "m_2": "85676623484277624682444820444702467499357910274165013418379125134075325543729",
            "v": "10113246594919968694243749150533421251903212516176280450944343564431109505938841891625247667364379412129405405348477095610002433568347856509159611335758818756926825063560770314873165029229176344268719893707027252437404047327923906866676449185116008460538437595639211714306227451873006607255100899512311105919649129622793316610076053733506986530767535638972131892478976764607887355989030338392858709274725683511041231799301175492600538984695657494046447608704545830499026706733632944905447307052632450973098625532585070622424882750416943328031030319205075858421333346786928868231555780815068857501287044930625867151895584283719258208379421233099516506594646921971059399550756016776860669930757562757559953692743338803382655236340538001870484617627827300880520094049323022699847759090147376675910494368188266848296015139094",
          },
          "r_credential": null,
        },
        "signature_correctness_proof": {
          "c": "71574462310595963848889785756072396101961922057659218227849729861834355800201",
          "se": "21478040510275170612337916082115329005924729160138728500731966689768864876350314127582065222070358453091326557476132640512915164393469162400241555486465031352127309617863096642685471579736279753483791845394120052052850579761097059767637641620637246275967251656241939449999728934398401723998520798896177945650388769630721777342623105417576378300802138205906595722648799399722998758345581952810879612170626417882503950639389717949927995386086595370764259106482228888541923217556032007348358437745103150644457074511256162173366561635465131422798361283669038048806849619337895126642478632186566293253597826848134434844979",
        },
        "values": {
          "test": {
            "encoded": "72155939486846849509759369733266486982821795810448245423168957390607644363272",
            "raw": "test",
          },
        },
        "witness": null,
      },
      "credentialId": "e1548638-fbe6-4e26-96fc-d536d453b380",
      "id": "9ec1f4c6-6b58-4c54-84d0-2fa74f83f70f",
      "linkSecretId": "ae343b1d-e2af-4706-9aae-2010a7f2c882",
      "metadata": {},
      "methodName": "indy",
      "updatedAt": "2024-02-28T09:36:42.994Z",
    },
  },
  "STORAGE_VERSION_RECORD_ID": {
    "id": "STORAGE_VERSION_RECORD_ID",
    "tags": {},
    "type": "StorageVersionRecord",
    "value": {
      "createdAt": "2023-03-18T18:53:44.041Z",
      "id": "STORAGE_VERSION_RECORD_ID",
      "metadata": {},
      "storageVersion": "0.5",
      "updatedAt": "2024-02-28T22:50:20.522Z",
    },
  },
}
`;
